File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int maxlen=0;
        int lastFruit=-1, secondLastFruit=-1;
        int currCount =0, lastFruitStreak=0;
        //treansverese through the fruits
        for(int fruit: fruits){
            //if fruit matches last two , expand the window
            if(fruit ==lastFruit || fruit ==secondLastFruit){
                currCount++;
            }else{
                //Reset Window size to streak+1
                currCount= lastFruitStreak+1;
            }
            //Update LastFruitStreak and fruit types
            if (fruit==lastFruit){
                lastFruitStreak++;
            }else{
                lastFruitStreak=1;
                secondLastFruit= lastFruit;
                lastFruit= fruit;
            }

            //update max window size
            maxlen = Math.max(maxlen, currCount);
        }
        return maxlen;
        
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution(object):
    def totalFruit(self, fruits):
        """
        :type fruits: List[int]
        :rtype: int
        """
        maxlen =0
        lastfruit = secondlastfruit= -1
        currcount = lastfruitstreak =0
        #Traverse through fruit array
        for fruit in fruits:
            #if currFruit matches last two, increase the window size
            if fruit == lastfruit or fruit == secondlastfruit:
                currcount +=1
            else:
                #Otherwise reset to streka size
                currcount = lastfruitstreak +1
            # Update last fruit streak and tracking
            if fruit == lastfruit:
                lastfruitstreak+=1
            else:
                lastfruitstreak=1
                secondlastfruit = lastfruit
                lastfruit = fruit
            maxlen = max(maxlen, currcount)
        return maxlen
        